Brick-Studios/BrickEngine
# ECS parent-child, isTrigger and input delaying

# Description
Creating a parent-child feature seemed easy, but it effected basically every single other system/feature. To prevent the massive scope-creep we were experiencing while building this PR we created [this document](https://docs.google.com/document/d/1nCuaUNFxBL2JVUqoIBVhkNmwr03OqiUPU62A5dUL8O0/edit?usp=sharing) to prevent ourselves from spending massive amounts of time on scenario's that would never come up in our game. Creating a full fledged parent-child system would have probably taken 3x as much time.

## Parent-Child
This PR centers around the parent-child feature in the EntityManager. An entity can now have a single parent and a parent can have multiple children. The transform of a child is always relative to that of it's parent, so when you need to know where an entity actually is you now need to use _getAbsoluteTransform_ in the EntityManager. When creating or setting the parent of an entity you can specify whether the current transform is already relative to its parent or not, this makes the system more usable and prevents bugs.

## CollisionDetectorInspector and isTrigger
To be able to detect that a collider is inside another collider, isTrigger was created by Jan. A collider now has the attribute isTrigger. The individual systems can then choose individually how to use the isTrigger. For instance the PhysicsSystem will not let an entity pass through an entity that has isTrigger and the PickupSystem will use isTrigger to detect whether the player is standing on a weapon.
Colliding or triggering with someone in your family hierarcy is impossible.

## Flipping
Before this PR we used to only have a flip in the Texture, this "flipping" has now been fully implemented into a directional feature. Every transform now has a xDirection and yDirection that can be positive or negative. The RenderingSystem will use the direction in Texture to flip the SDL_Texture accordingly.

The PhysicsComponent now has a flipX and flipY, these are used to determine what the PhysicsSystem should do when a flip should or could occur.

```
// If the entity is a parent, only the direction will get flipped
// If the entity is a child, both the direction and position get flipped
bool flipX;
bool flipY;
```

The position flipping relates to whether should have its position reversed when the parent gets flipped. A good example of this is the position of a weapon relative to its player that will get flipped when the player flips left or right.

# Kinematic
First we had `bool kinematic` now be have `Kinematic::SHOULD_NOT_BE_KINEMATIC`, beautiful is it not? From our features that needed to be supported in parent-child, we devised that a child should **always** be kinematic. This is something we try to enforce when a parent is set or when the child is thrown out the parents house. But to do this we had to remember what the original kinematic state was before the setting of a parent. So we now have:
```
// When using this enum in a system, use kinematic != IS_KINEMATIC or kinematic == IS_KINEMATIC
// to automaticly include or exclude the SHOULD_NOT_BE_KINEMATIC value
enum class Kinematic {
    IS_KINEMATIC,
    // This value should act the same as IS_KINEMATIC, however when a child is released
    // this value will get changed back to the original IS_NOT_KINAMTIC value
    SHOULD_NOT_BE_KINEMATIC,
    IS_NOT_KINEMATIC
};
```

# How can this be tested?

[Check the PR in BeastArena](https://github.com/Brick-Studios/BeastArena/pull/28).

# Checklist:
